File: src/com/dixie/adventofcode/aoc2017/Day3.java
package com.dixie.adventofcode.aoc2017;

import com.dixie.adventofcode.lib.Day;
import com.dixie.adventofcode.lib.Direction;
import com.dixie.adventofcode.lib.Space2D;

import java.awt.*;
import java.util.Arrays;
import java.util.List;

public class Day3 extends Day {
  public static void main(String[] args) {
    new Day3().solve();
  }

  private Point[] spiral;

  @Override
  protected void prepare(List<String> lines) {
    int target = Integer.parseInt(lines.get(0));
    spiral = new Point[target + 1];
    Point p = new Point(0, 0);
    spiral[1] = p;
    Direction dir = Direction.RIGHT;
    int len = 1;
    int line = 0;
    int num = 2;
    while (num <= target) {
      for (int l = 0; l < len && num <= target; l++) {
        p = dir.apply(p);
        spiral[num++] = p;
      }
      dir = dir.turnLeft();
      line++;
      if (line % 2 == 0) len++;
    }
  }

  @Override
  protected Object part1(List<String> lines) {
    Point last = spiral[spiral.length - 1];
    return Math.abs(last.x) + Math.abs(last.y);
  }

  @Override
  protected Object part2(List<String> lines) {
    Space2D<Integer> sumSpiral = new Space2D<>();
    sumSpiral.setValueAt(new Point(0, 0), 1);
    for (int i = 2; i < spiral.length; i++) {
      Point p = spiral[i];
      int sum = Arrays.stream(Direction.values())
          .map(dir -> dir.apply(p))
          .mapToInt(c -> sumSpiral.getValueAt(c, 0))
          .sum();
      sumSpiral.setValueAt(p, sum);
      if (sum >= spiral.length) {
        return sum;
      }
    }
    return 0;
  }
}

File: src/com/dixie/adventofcode/aoc2023/Day3.kt
package com.dixie.adventofcode.aoc2023

import com.dixie.adventofcode.klib.Day
import com.dixie.adventofcode.lib.Direction
import java.awt.Point

class Day3 : Day() {
  private lateinit var numbers: Map<Point, Number>
  private lateinit var specials: List<Point>

  override fun prepare() {
    val numbers = mutableMapOf<Point, Number>()
    val specials = mutableListOf<Point>()
    for (y in lines.indices) {
      val line = lines[y]
      val digits = mutableListOf<Char>()
      val locations = mutableListOf<Point>()
      for (x in line.indices) {
        val ch = line[x]
        val loc = Point(x, y)
        if (ch.isDigit()) {
          digits += line[x]
          locations += loc
        }
        if (x == line.lastIndex || !ch.isDigit()) {
          if (!ch.isDigit() && ch != '.') specials += loc
          if (locations.isNotEmpty()) {
            val num = Number(locations[0], digits.joinToString("").toInt())
            locations.forEach { numbers[it] = num }
            digits.clear()
            locations.clear()
          }
        }
      }
    }
    this.numbers = numbers
    this.specials = specials
  }

  override fun part1(): Any {
    val partNumbers = mutableSetOf<Number>()
    specials.forEach { p ->
      for (neighbour in neighbours(p)) {
        numbers[neighbour]?.let(partNumbers::add)
      }
    }
    return partNumbers.sumOf(Number::value)
  }

  override fun part2() =
    specials
      .asSequence()
      .filter { p -> lines[p.y][p.x] == '*' }
      .map { neighbours(it).mapNotNull(numbers::get).toSet().toList() }
      .filter { it.size == 2 }
      .sumOf { it[0].value * it[1].value }

  private fun neighbours(point: Point) = Direction.values().map { it.apply(point) }

  data class Number(val start: Point, val value: Int)
}

fun main() {
  Day3().run()
}
